### ISSUE TITLE
Coverage build fails

### ISSUE THREAD

**Issue by Pamplemousse** [google/oss-fuzz#6385]
The documentation on how to [build fuzz targets for coverage](https://google.github.io/oss-fuzz/advanced-topics/code-coverage/#build-fuzz-targets) seems to imply that running `python infra/helper.py build_fuzzers --sanitizer=coverage $PROJECT_NAME` should compile a target for coverage.

When run as is, the build fails, with a lot of `ld` errors about `undefined reference to...`
Here are a couple:

```shell
[...]
/nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/std_function.h:(.text._ZNSt17_Function_handlerIFbvEZN3nix10ThreadPool6doWorkEbE3$_0E10_M_managerERSt9_Any_dataRKS5_St18_Manager_operation[_ZNSt17_Function_handlerIFbvEZN3nix10ThreadPool6doWorkEbE3$_0E10_M_managerERSt9_Any_dataRKS5_St18_Manager_operation$d2c559eb12798b5444b9d144492f93e1]+0x8b): undefined reference to `__sanitizer_cov_trace_pc'
/nix/store/sgrwh6jsrshrnicqi67fck336q6ldh2d-binutils-2.35.1/bin/ld: /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/std_function.h:(.text._ZNSt17_Function_handlerIFbvEZN3nix10ThreadPool6doWorkEbE3$_0E10_M_managerERSt9_Any_dataRKS5_St18_Manager_operation[_ZNSt17_Function_handlerIFbvEZN3nix10ThreadPool6doWorkEbE3$_0E10_M_managerERSt9_Any_dataRKS5_St18_Manager_operation$d2c559eb12798b5444b9d144492f93e1]+0x97): undefined reference to `__sanitizer_cov_trace_pc_guard'
/nix/store/sgrwh6jsrshrnicqi67fck336q6ldh2d-binutils-2.35.1/bin/ld: src/libutil/thread-pool.o: in function `std::_Function_base::_Base_manager<nix::ThreadPool::doWork(bool)::$_0>::_M_manager(std::_Any_data&, std::_Any_data const&, std::_Manager_operation)':
/nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/std_function.h:184: undefined reference to `__sanitizer_cov_trace_pc'
/nix/store/sgrwh6jsrshrnicqi67fck336q6ldh2d-binutils-2.35.1/bin/ld: /nix/store/5qjycalzb9sqzvqg65kf5zimqwjabm9g-gcc-10.3.0/include/c++/10.3.0/bits/std_function.h:184: undefined reference to `__sanitizer_cov_trace_pc_guard'
```

This is not too surprising as there is nothing in my `build.sh` setting any flags in case `$SANITIZER` equals `coverage` (and nothing seems to automagically sets flags).

I looked through the documentation, as well as in other projects code, issues, in the search of hints about how (and what - as LLVM proposes two ways of instrumenting for coverage) coverage instrumentation is supposed to be used, and I still have no clue...

Are there any environment variables set with coverage flags?
Is there a more detailed documentation listing those?

**Comment by DavidKorczynski** [comment]
No environment variables are set out of the ordinary `CFLAGS`, `CXXFLAGS` and `LIB_FUZZING_ENGINE` - as long as you use those then you should be fine. Also, remember to use `CC` and `CXX` for compiling and linking

**Comment by Pamplemousse** [comment]
> then you should be fine

Well, it is not, hence the issue in the first place...

When I make my `build.sh` to simply `echo` those three variables, and calL `python infra/helper.py build_fuzzers --sanitize=coverage nix`, here is what I get:
```
# CXXFLAGS
-fprofile-instr-generate -fcoverage-mapping -pthread -Wl,--no-as-needed -Wl,-ldl -Wl,-lm -Wno-unused-command-line-argument -stdlib=libc++
-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fprofile-instr-generate -fcoverage-mapping -pthread -Wl,--no-as-needed -Wl,-ldl -Wl,-lm -Wno-unused-command-line-argument -stdlib=libc++
# CFLAGS
-O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fprofile-instr-generate -fcoverage-mapping -pthread -Wl,--no-as-needed -Wl,-ldl -Wl,-lm -Wno-unused-command-line-argument
# LIB_FUZZING_ENGINE
-fsanitize=fuzzer
```

Interestingly, in the shell given by `python infra/helper.py build_fuzzers --sanitize=coverage nix`, the instrumentation flags are absent...

**Comment by Pamplemousse** [comment]
My guess now is that it's not fine because I am missing flags to pass to `ld` :facepalm:

**Comment by Pamplemousse** [comment]
I guess I should be forwarding those flags to the linking of the shared libraries I compile... :facepalm:

**Comment by DavidKorczynski** [comment]
> Interestingly, in the shell given by `python infra/helper.py build_fuzzers --sanitize=coverage nix`, the instrumentation flags are absent...

That command won't produce a shell.

> I guess I should be forwarding those flags to the linking of the shared libraries I compile... facepalm

Yep, or at least you should make sure the flags used for compilation match with those for linking.

**Comment by Pamplemousse** [comment]
> That command won't produce a shell.

My bad, wrong copy/paste; I meant `python3 infra/helper.py shell nix --sanitizer=coverage`.

**Comment by Pamplemousse** [comment]
It seems that adding the flag `-fsanitize=fuzzer-no-link` for compiling a shared library conflicts with the provided coverage flags...

**Comment by DavidKorczynski** [comment]
Did you ever resolve this one @Pamplemousse ? Am not entirely sure if you're suggesting the flags provided by OSS-Fuzz are incompatible for coverage builds?

**Comment by MarcoFalke** [comment]
Is this still an issue?

**Comment by Pamplemousse** [comment]
@DavidKorczynski Yes they are, when `-fsanitize=fuzzer-no-link` is used for libraries. "Current" (haven't worked on OSS-fuzz integration for a while) solution is to remove the `-fsanitize=fuzzer-no-link` flag from the libraries I compile when building for coverage: https://github.com/google/oss-fuzz/pull/6317/files#diff-ea3897e44e696bd00c3430ca2ba03e6521eb8af280d2b2d3dfdc14751564eca6R47-R50 .

@MarcoFalke I don't know as I haven't worked with OSS-fuzz roughly since the opening of this issue.
Also, I am not too sure if `-fsanitize=fuzzer-no-link` in libraries is that important when building for coverage.
